### PR TITLE
Data: Add rule indexing interfaces and xml index

### PR DESCRIPTION
**Pull request by HolonProduction**
Implements parts of #8 #9 #11 #12

### CODE REVIEW

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/xml/JqaXmlRuleIndexingStrategy.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+) : JqaRuleIndexingStrategy {
+    class Factory : JqaRuleIndexingStrategyFactory {
+        override fun create(project: Project): JqaXmlRuleIndexingStrategy = JqaXmlRuleIndexingStrategy(project)
+    }
+
+    // TODO: Use a separate type based index and support the [type] filter.
+    override fun getAll(type: JqaRuleType): List<JqaRuleDefinition> {
+        return FileBasedIndex.getInstance().getAllKeys(NameIndex.Util.NAME, project).map {
+            ValueBasedJqaRuleDefinition(
+                it,
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/xml/JqaXmlRuleIndexingStrategy.kt:25]: nit: naming

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/xml/JqaXmlRuleIndexingStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                    } else {
+                        res.set(
+                            ValueBasedJqaRuleDefinition(
+                                name,
+                                JqaRuleType.CONCEPT, // TODO: Extract correct type
+                                psiElement,
+                            )
+                        )
+                        false
+                    }
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/xml/JqaXmlRuleIndexingStrategy.kt:57]:
```suggestion
                    if (psiElement == null) return@processValues true
                    
                    res.set(
                        ValueBasedJqaRuleDefinition(
                            name,
                            JqaRuleType.CONCEPT, // TODO: Extract correct type
                            psiElement,
                        )
                    )
                    false
```

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/JqaRuleIndexingService.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    override fun extensionRemoved(extension: JqaRuleIndexingStrategyFactory, pluginDescriptor: PluginDescriptor) {
+        indexes.clear()
+        for (factory in JqaRuleIndexingStrategyFactory.Util.EXTENSION_POINT.extensions) {
+            indexes.add(factory.create(project))
+        }
+    }
+
+    fun getAll(type: JqaRuleType): List<JqaRuleDefinition> = indexes.flatMap { it.getAll(type) }
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/JqaRuleIndexingService.kt:43]: Nit: Naming, for exmaple `getAllByType()`

**Review by A-K-O-R-A — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/xml/NameIndex.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import com.intellij.util.io.EnumeratorIntegerDescriptor
+import com.intellij.util.io.EnumeratorStringDescriptor
+import com.intellij.util.io.KeyDescriptor
+import com.intellij.util.xml.DomManager
+
+/**
+ * Indexes all rules, and associates their offset in the file to their name.
+ */
+class NameIndex : FileBasedIndexExtension<String, Int>() {
+    object Util {
```
[the remark this answers is not in the document]
Comment (HolonProduction) [reply src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/xml/NameIndex.kt:16]: Will keep it at `Util` for now. Background: Those stuff would normally go in a companion object, but due to IntelliJ Platform restrictions we need a named object. To make this usable across the projekt we should name those objects that we use in place of companions consistently. And `Util` is what the quick fix suggests so it is somewhat universal.

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/JqaRuleIndexingService.kt (hunk 3 above)
Comment (HolonProduction) [reply src/main/kotlin/org/jqassistant/tooling/intellij/plugin/data/rules/JqaRuleIndexingService.kt:43]: Those methods will be reworked when we switch to MultiReferences, will adjust this then.

**Review by A-K-O-R-A — APPROVED** [review]

**Review by Valvon — APPROVED** [review]

**Pull request merged**